### PR TITLE
feat: support data transfer protection in HDFS artifacts (#13482)

### LINKED ISSUE #13482: Support data transfer protection in HDFS artifacts
# Summary

What change needs making?

**AS-IS:**
An error occurs when connecting to an HDFS cluster with `dfs.data.transfer.protection` enabled. The error message is `read: connection reset by peer`, similar to the issue described in [this issue](https://github.com/colinmarc/hdfs/issues/184).

**TO-BE:**
The HDFS client supports the data transfer protection option, which can be configured to `authentication`, `integrity`, `privacy`, or left empty, as detailed in [this pull request](https://github.com/colinmarc/hdfs/pull/236). With this configuration, the issue described above is resolved.

This feature is available in versions v2.4.0, v2.3.0, v2.2.1, and v2.2.0 (see [commit](https://github.com/colinmarc/hdfs/commit/1596ee1307eb423a2ed092713f1368157d67ce86)). Since we are using version v2.4.0 of the HDFS client (as specified [here](https://github.com/argoproj/argo-workflows/blob/main/go.mod#L18)), we can utilize this feature.

## Use Cases

When would you use this?

As shown in https://github.com/argoproj/argo-workflows/blob/main/examples/hdfs-artifact.yaml, I am working on storing artifacts in HDFS. In my environment, the HDFS cluster has `dfs.data.transfer.protection` set to `authentication`. When I run the example, I encounter a `read: connection reset by peer` error.

However, with the enhancement, the operation would succeed by properly configuring the `dfs.data.transfer.protection` setting.

---

**Message from the maintainers**:

Love this feature request? Give it a 👍. We prioritise the proposals with the most 👍.

### PR DESCRIPTION
**Pull request by Looka149**
Fixes #13482

### Motivation

As shown in https://github.com/argoproj/argo-workflows/blob/main/examples/hdfs-artifact.yaml, I am working on storing artifacts in HDFS. In my environment, the HDFS cluster has `dfs.data.transfer.protection` set to `authentication`. When I run the example, I encounter a `read: connection reset by peer` error.

### Modifications

The HDFS client supports the data transfer protection option, which can be configured to `authentication`, `integrity`, `privacy`, or left empty, as detailed in [this pull request](https://github.com/colinmarc/hdfs/pull/236). With this configuration, the issue described above is resolved.

This feature is available in versions v2.4.0, v2.3.0, v2.2.1, and v2.2.0 (see [commit](https://github.com/colinmarc/hdfs/commit/1596ee1307eb423a2ed092713f1368157d67ce86)). Since we are using version v2.4.0 of the HDFS client (as specified [here](https://github.com/argoproj/argo-workflows/blob/main/go.mod#L18)), we can utilize this feature.

### Verification

Referring to the [hdfs-artifact.yaml example](https://github.com/argoproj/argo-workflows/blob/main/examples/hdfs-artifact.yaml), you can specify the value of `artifacts.hdfs.dataTransferProtection` as 'authentication', 'integrity', 'privacy', or leave it empty.

```yaml
  - name: print-message
    inputs:
      artifacts:
      - name: message
        path: /tmp/message
        hdfs:
          dataTransferProtection: authentication
```

If an incorrect value is provided, the following error will occur.

```bash
$ kubectl get workflow
NAME                  STATUS   AGE     MESSAGE
hdfs-artifact-td82h   Failed   33m     invalid spec: templates.artifact-example.steps[1].consume-artifact templates.print-message.inputs.artifacts.message.hdfs.dataTransferProtection must be one of `authentication`, `integrity`, `privacy`, or empty
```

### CODE REVIEW

**Comment by Looka149** [comment]
@terrytangyuan @sarabala1979 
Hi, is there anything else needed to proceed with the approval? Thanks.

File: workflow/artifacts/hdfs/hdfs.go [hunk 1]
```diff
@@ -73,6 +80,14 @@ func ValidateArtifact(errPrefix string, art *wfv1.HDFSArtifact) error {
 	if hasKrbCCache && (art.KrbServicePrincipalName == "" || art.KrbConfigConfigMap == nil) {
 		return errors.Errorf(errors.CodeBadRequest, "%s.krbServicePrincipalName and %s.krbConfigConfigMap are required with %s.krbCCacheSecret", errPrefix, errPrefix, errPrefix)
 	}
+
+	switch art.DataTransferProtection {
+	case "", DataTransferProtectionAuthentication, DataTransferProtectionIntegrity, DataTransferProtectionPrivacy:
+		// Do nothing
+	default:
+		return errors.Errorf(errors.CodeBadRequest, "%s.dataTransferProtection must be one of 'authentication', 'integrity', 'privacy', or empty", errPrefix)
+	}
```
Comment (juliev0) [inline workflow/artifacts/hdfs/hdfs.go:89]: Thanks for adding the validation here, although I'm torn on whether we should validate this or not. On one hand, if the `ClientOptions` ever change for `DataTransferProtection` then we'd have to update this when the version in go.mod gets updated.

**Comment by juliev0** [comment]
@Looka149 Your PR looks good. I can merge it. Only thing I'm thinking is that it may be better not to validate on these specific strings which could theoretically change, per my [comment](https://github.com/argoproj/argo-workflows/pull/13483#discussion_r1742878461), but I may be convinced otherwise. Thanks.

File: workflow/artifacts/hdfs/hdfs.go (hunk 1 above)
Comment (Looka149) [reply workflow/artifacts/hdfs/hdfs.go]: I agree that your concern is valid. I initially added validation based on the three modes specified in https://github.com/colinmarc/hdfs/blob/master/client.go#L24-L28. While I don't expect these options to change frequently, **I agree that if they do change in the future, this validation might be easily overlooked.**

**Comment by Looka149** [comment]
@juliev0 Thank you for reviewing my PR, I agreed with your feedback and have removed the validation. Thank you.

**Comment by juliev0** [comment]
@Looka149 assuming the CI failures aren't related to your change, would you mind adding [empty commits ](https://www.freecodecamp.org/news/how-to-push-an-empty-commit-with-git/) until the CI passes?

**Comment by Looka149** [comment]
@juliev0 I have merged the main branch, and the CI tests have completed successfully.

**Review by juliev0 — APPROVED** [review]

File: workflow/artifacts/hdfs/util.go [hunk 2]
```diff
@@ -8,9 +8,10 @@ import (
 	"github.com/jcmturner/gokrb5/v8/config"
 )
 
-func createHDFSClient(addresses []string, user string, krbOptions *KrbOptions) (*hdfs.Client, error) {
+func createHDFSClient(addresses []string, user string, dataTransferProtection string, krbOptions *KrbOptions) (*hdfs.Client, error) {
 	options := hdfs.ClientOptions{
-		Addresses: addresses,
+		Addresses:              addresses,
+		DataTransferProtection: dataTransferProtection,
```
Comment (agilgur5) [inline workflow/artifacts/hdfs/util.go:14]: To make artifacts more sustainable (other than plugins #5862), I've been arguing we should perhaps accept arbitrary YAML string output. E.g. in this case:
```yaml
clientOptions: |
  dataTransferProtection: true
```

That way we don't have to add a new feature every time a user wants to configure something else

**Pull request merged**